File: 프로그래머스/1/42889. 실패율/실패율.py
def solution(N, stages):
    answer = []
    # 스테이지 길이
    total= len(stages)
    for i in range(1,N+1):
        # 각 스테이지 실패인원 count
        cnt = stages.count(i)
        try:
            tmp = cnt/total
            answer.append((i,tmp))
        except ZeroDivisionError:
            answer.append((i, cnt))
        total-= cnt
    # 실패율(x[1])을 기준으로 내림차순 정렬하고, 만약 실패율이 같을 경우 번호순으로 정렬(-x[0])
    answer.sort(reverse=True, key = lambda x : (x[1], -x[0]))
    
    result = []
    for i, fail in answer:
        result.append(i)
    return result

File: 프로그래머스/1/42889. 실패율/실패율.java
import java.util.*;

class Solution {
    public int[] solution(int N, int[] stages) {
        int[] answer = new int[N];
        double length = stages.length;
        ArrayList<Info> list = new ArrayList();
        
        for (int i = 1; i<=N; i++){ // stage 단계
            double cnt = 0;
            for (int j = 0; j < stages.length; j++){
                if (stages[j] == i)
                    cnt++;
            }
            if (cnt == 0)
                list.add(new Info(0, i));
            else
                list.add(new Info((cnt/length), i));
            length -= cnt;
        }
        
        Collections.sort(list);
        
        for (int i = 0; i < list.size(); i++)
            answer[i] = list.get(i).index;
        
        return answer;
    }
}

class Info implements Comparable<Info>{
    double failure;
    int index;
    
    public Info(double failure, int index){
        this.failure = failure;
        this.index = index;
    }
    
    public int compareTo(Info o){
        if (this.failure == o.failure)
            return this.index - o.index;
        return Double.compare(o.failure, this.failure);
    }
}